python/ray/data/datasource/parquet_datasource.py

```diff
@@ -77,51 +90,79 @@ def prepare_read(
+            del dataset_kwargs["partitioning"]
+        filesystem = _wrap_s3_serialization_workaround(filesystem)
 
-            # Ensure that we're reading at least one dataset fragment.
-            assert len(pieces) > 0
+        def read_files(
+            files: List[PartitionedFile], filesystem, **dataset_kwargs
+        ) -> Iterator[pa.Table]:
+            if isinstance(filesystem, _S3FileSystemWrapper):
+                filesystem = filesystem.unwrap()
 
-            from pyarrow.dataset import _get_partition_keys
+            # Ensure that we're reading at least one file.
+            assert len(files) > 0
 
             ctx = DatasetContext.get_current()
             output_buffer = BlockOutputBuffer(
                 block_udf=_block_udf, target_max_block_size=ctx.target_max_block_size
             )
 
-            logger.debug(f"Reading {len(pieces)} parquet pieces")
+            logger.debug(f"Reading {len(files)} parquet files")
             use_threads = reader_args.pop("use_threads", False)
-            for piece in pieces:
-                part = _get_partition_keys(piece.partition_expression)
-                batches = piece.to_batches(
-                    use_threads=use_threads,
-                    columns=columns,
-                    schema=schema,
-                    batch_size=PARQUET_READER_ROW_BATCH_SIZE,
-                    **reader_args,
+            for f in files:
+                ds = pq.ParquetDataset(
+                    f.path,
+                    **dataset_kwargs,
+                    filesystem=filesystem,
+                    use_legacy_dataset=False,
+                    split_row_groups=False,
                 )
-                for batch in batches:
-                    table = pyarrow.Table.from_batches([batch], schema=schema)
-                    if part:
-                        for col, value in part.items():
-                            table = table.set_column(
-                                table.schema.get_field_index(col),
-                                col,
-                                pa.array([value] * len(table)),
+                for fragment in ds.fragments:
+                    batches = fragment.to_batches(
+                        use_threads=use_threads,
+                        columns=columns,
+                        schema=schema,
+                        batch_size=PARQUET_READER_ROW_BATCH_SIZE,
+                        **reader_args,
+                    )
+                    for batch in batches:
+                        table = pa.Table.from_batches([batch], schema=schema)
+                        # TODO(ekl) we have to manually apply the filter on partitioned
+                        # columns after the read. This is an artifact of the way we
+                        # work around file fragment serialization issues.
+                        if f.partition_keys and "filter" in reader_args:
+                            raise NotImplementedError(
```

Comment:
Unfortunately, this is non-trivial to implement. Consider punting on this for now, since metadata performance is more important than filter support.

(could consider as a P1 for GA)